core/src/main/java/org/eclipse/hono/util/AddressHelper.java

```diff
@@ -0,0 +1,96 @@
+     * @return The target address.
+     * @throws NullPointerException if endpoint or tenant is {@code null}.
+     */
+    public static String getTargetAddress(final String endpoint, final String tenantId, final String resourceId, final String addressRewriteRule) {
+        final StringBuilder addressBuilder = new StringBuilder(Objects.requireNonNull(endpoint))
+                .append("/").append(Objects.requireNonNull(tenantId));
+        if (!Strings.isNullOrEmpty(resourceId)) {
+            addressBuilder.append("/").append(resourceId);
+        }
+
+        return AddressHelper.rewrite(endpoint, addressBuilder.toString(), addressRewriteRule);
+    }
+
+    /**
+     * Rewrites address for a certain endpoint according to the provided rule.
+     *
+     * @param endpoint The messaging endpoint.
+     * @param address The address to be rewritten.
+     * @param rewriteRule The rewrite rule. The rule is in the <em>$PATTERN $SUBSTITUTION</em> format.
+     *                    The <em>$PATTERN</em> should match the original address. It uses the regular Java pattern syntax. There
+     *                    are two helper constants introduced:
+     *                    <ul>
+     *                      <li><em>ENDPOINT</em> - that will match the current endpoint being used</li>
+     *                      <li><em>ID</em> - a pattern for id representations in the system</li>
+     *                    </ul>
+     *
+     *                    Example:
+     *                    <em>(${ENDPOINT})/(${ID}) test-vhost/$1/$2</em>
+     *                    will rewrite
+     *                    <em>telemetry/DEFAULT_TENANT</em> to <em>test-vhost/telemetry/DEFAULT-TENANT</em>
+     *                    If the rule is {@code null}, in the wrong format or the pattern doesn't match the address,
+     *                    the original address will be returned.
+     * @return The address to be used.
+     */
+    public static String rewrite(final String endpoint, final String address, final String rewriteRule) {
+        if (Strings.isNullOrEmpty(rewriteRule)) {
+            return address;
+        }
+
+        final String[] elements = rewriteRule.split(" ", 2);
+        if (elements.length != 2) {
+            return address;
+        }
+        return rewrite(endpoint, address, elements[0], elements[1]);
+    }
+
+    static String rewrite(final String endpoint, final String address, final String regex, final String replacement) {
+        try {
+            return address.replaceAll(regexSubstitues(endpoint, regex), replacement);
+        } catch (PatternSyntaxException pe) {
+            return address;
+        }
+    }
+
+    static String regexSubstitues(final String endpoint, final String regex) {
+        return regex
+                .replaceAll("\\$\\{ENDPOINT\\}", endpoint)
+                .replaceAll("\\$\\{ID\\}", RegistryManagementConstants.DEFAULT_ID_REGEX);
```

Comment:
what if the tenant ID and/or device ID regex are being configured explicitly, i.e. do not have their default value?

Reply:
It won't work. But it's just a simple helper, so it's not something that you must use to make it work. I just thought some these would make configuration look simpler for common use-cases.

We can do more in this area and do substitutions based on configuration, environment variables, etc. But I didn't want to go there until we have the basics sorted.

If this is confusing we can remove substitutions for the time being and rethink them later on.

Reply:
I guess it will be very hard to trace and understand the source of the problem in this case. So FMPOV we should either omit the substitution logic for the time being or integrate it properly with the configurable tenant and device ID patterns ...

Reply:
OK, I'll remove it completely for now.